java/ql/src/semmle/code/java/dataflow/internal/TaintTrackingUtil.qll

```diff
@@ -680,3 +678,14 @@ private MethodAccess callReturningSameType(Expr ref) {
   ref = result.getQualifier() and
   result.getMethod().getReturnType() = ref.getType()
 }
+
+private RefType entrypointType() {
+  result = any(RemoteFlowSource s | exists(Parameter p | p = s.asParameter())).getType() or
+  result = entrypointType().getASubtype() or
+  result = entrypointType().getAField().getType()
+}
+
+private predicate entrypointFieldStep(DataFlow::Node src, DataFlow::Node sink) {
+  src = DataFlow::getFieldQualifier(sink.asExpr().(FieldRead)) and
+  src.getType() = entrypointType()
```

Comment:
This logic and the entirety of `entrypointType()` looks reasonable, but I'm not sure it works too well with generics. We should probably rephrase this logic to work in terms of source declarations. I.e. the `getType()`s should be `getType().getSourceDeclaration()`, the declared result type of `entrypointType()` should be `SrcRefType` and `result = entrypointType().getASubtype()` should instead be `result.getASourceSupertype() = entrypointType()`.

Reply:
Good catch! Shouldn't it be something like this though?

```ql
private SrcRefType entrypointType() {
  result =
    any(RemoteFlowSource s | s instanceof DataFlow::ExplicitParameterNode)
        .getType()
        .(RefType)
        .getSourceDeclaration() or
  result.getASourceSupertype() = entrypointType() or
  result = entrypointType().getAField().getType().(RefType).getSourceDeclaration()
}

private predicate entrypointFieldStep(DataFlow::Node src, DataFlow::Node sink) {
  src = DataFlow::getFieldQualifier(sink.asExpr().(FieldRead)) and
  src.getType().(RefType).getSourceDeclaration() = entrypointType()
}
```

i.e. basically source types need to be used everywhere for the comparison to work. Or am I misunderstanding something?

Reply:
No misunderstandings. That implementation is exactly what I meant.

Reply:
Amazing, thank you! Fixed in 8250cec